### PR TITLE
【new feature 3】Update gas for transfer mnt precompiled

### PR DESCRIPTION
**Pull request by 516108736**
https://github.com/QuarkChain/pyquarkchain/pull/854

### CODE REVIEW

File: core/vm/contracts.go [hunk 1]
```diff
@@ -540,38 +542,81 @@ func (c *transferMnt) Run(input []byte, evm *EVM, contract *Contract) ([]byte, e
...
-	data := getData(input, 96, uint64(len(input)-96))
 	t := evm.TransferTokenID
-	evm.TransferTokenID = mnt.Uint64()
-	ret, remainedGas, err := evm.Call(contract.caller, toAddr, data, contract.Gas, value)
+	evm.TransferTokenID = tokenID.Uint64()
+	gasToCall := contract.Gas - gasCost
+	if value.Sign() > 0 {
+		gasToCall += params.CallStipend
+	}
+	contract.UseGas(gasCost)
```
Comment (pingke) [inline core/vm/contracts.go]: same with pyquarkchain?

Comment (516108736) [reply core/vm/contracts.go:595]: @syntrust please double check it

File: core/vm/contracts.go [hunk 2]
```diff
@@ -540,38 +542,62 @@ func (c *transferMnt) Run(input []byte, evm *EVM, contract *Contract) ([]byte, e
...
-	if !evm.StateDB.Exist(toAddr) && big.NewInt(0).Cmp(value) < 0 {
-		if !contract.UseGas(params.CallNewAccountGas) {
-			contract.Gas = 0
-			return nil, ErrOutOfGas
-		}
+	data := getData(input, 96, uint64(len(input)-96))
+
+	// Token ID should be within range
+	if tokenID.Uint64() > qCommon.TOKENIDMAX {
+		return nil, fmt.Errorf("tokenid %v > TOKENIDMAX %v", tokenID, qCommon.TOKENIDMAX)
```
Comment (syntrust) [inline core/vm/contracts.go:562]: should use up all gas here

File: core/vm/contracts.go [hunk 3]
```diff
@@ -540,38 +542,62 @@ func (c *transferMnt) Run(input []byte, evm *EVM, contract *Contract) ([]byte, e
...
 	}
-	if big.NewInt(0).Cmp(value) < 0 {
-		if !contract.UseGas(params.CallValueTransferGas) {
-			contract.Gas = 0
-			return nil, ErrOutOfGas
+
+	// Doesn't allow target address to be this precompiled contract itself
+	if toAddr.String() == common.HexToAddress(transferMntAddr).String() {
+		return nil, fmt.Errorf("re call toAddr %v", toAddr.String())
```
Comment (syntrust) [inline core/vm/contracts.go:567]: use all gas

File: core/vm/contracts.go [hunk 4]
```diff
@@ -540,38 +542,62 @@ func (c *transferMnt) Run(input []byte, evm *EVM, contract *Contract) ([]byte, e
...
+		gasCost += params.CallValueTransferGas
+		if !evm.StateDB.Exist(toAddr) {
+			gasCost += params.CallNewAccountGas
 		}
 	}
 
-	data := getData(input, 96, uint64(len(input)-96))
+	//Out of gas
+	if contract.Gas < gasCost {
+		return nil, fmt.Errorf("run transferMnt failed contracr.gas %v < gasCost %v", contract.Gas, gasCost)
```
Comment (syntrust) [inline core/vm/contracts.go:580]: use all gas

**Pull request merged**